### PR TITLE
Fix for #288 to keep the 'Connection to Server' popup from appearing …

### PR DESCRIPTION
**Pull request by carstendev**
…in the Get Blank Form after the device orientation changes.

The <tt>downloadFormList()</tt> method now only gets executed, when there are no entries in the <tt>mFormNamesAndUrls</tt> map and <tt>getLastNonConfigurationInstance()</tt> method returns null.

This is necessary because the <tt>mDownloadFormListTask</tt> gets set to null in the <tt>formListDownloadingComplete()</tt> method, after the <tt>mDownloadFormListTask</tt> is executed. Therefore, the <tt>getLastNonConfigurationInstance()</tt> method always returns null, for an instance of <tt>mDownloadFormListTask</tt>, after a full initialization cycle of the activity.<br>
While trying to figure out, what was causing the problem, I found it a lot harder than expected, to find the cause. I believe, that this was in part, because of the class design, which I found very hard to follow, because of the many side effects and responsibilities the class has.<br>
I think implementing this change, is like a drop on hot a stone. @lognaturel Maybe this class should be refactored because I only see it getting worse in the future, if not addressed.

### CODE REVIEW

**Comment by lognaturel** [comment]
Thanks so much for taking the time to look at this @carstendev.

You're completely right that there's a big need for refactoring and this bug really highlights it. A lot of Collect was written as Android was being developed so the original authors were without the benefit of best practices. I don't know how familiar you are with the project or its history but it was housed at the University of Washington and developed by students for a long time and is just now being transitioned to a true community project. Getting some fresh eyes like yours on the code will greatly improve its quality!

If you had a moment to file an issue with some general refactoring recommendations, that would be very helpful. Having that issue will give anyone interested a chance to comment and help set refactoring priorities. A lot of the challenges you saw in this class are present across the codebase so I think it makes sense to look at them from a project level.

As you've probably noticed, the project also lacks automated tests. This makes refactoring higher-risk so we may want to consider a testing plan before diving into refactoring. If you have expertise and interest in testing, please consider filing an issue with testing recommendations too to get the conversation started! This could be around tools, most important code paths to test, etc. I don't believe many of the current contributors have much testing experience so that would be an important contribution.

I'll look at this class again in a bit more depth tomorrow but I think your simple fix is a good "bandaid" for the time being and that the refactor would best be done as part of a project-wide initiative.

**Comment by lognaturel** [comment]
This is a definite improvement so it makes sense to merge now as we figure out a longer-term solution to addressing and avoiding similar bugs.

**Comment by carstendev** [comment]
@lognaturel Unfortunately, I don´t have any special knowledge when it comes to testing, but I will definitely file an issue and try to point out some of the problems I´ve seen in the codebase, when I get back from a trip for work. I´ll also try to suggest some solutions in the issue.

**Comment by lognaturel** [comment]
Great, have a nice trip, @carstendev! Looking forward to starting the discussion around refactoring when you get back.
